PluginCode/src/main/java/smr/cs/ualberta/libcomp/action/ReplacementAction.java

```diff
@@ -932,18 +606,181 @@ public void detectImports(@NotNull final PsiFile psiFile, @NotNull final Editor
+                        depObj.setEnableddomain(false);
+                    }
+                    else {
+                        depObj.setEnableddomain(true);
+                    }
+
+                    gradleDepList.add(depObj);
+
+                    if (depObj.getEnableddomain()) {
+                        editorModel.addLineHighlighter(currentLine,
+                                DebuggerColors.BREAKPOINT_HIGHLIGHTER_LAYER + 1, softerAttributes);
 
-         try {
-             detectImports(psiFile, editor, project_name);
-         }
-         catch(Exception e) {
-             e.printStackTrace();
-         }
+                    }
+                }
+            }
+            ++currentLine;
+            boolean isContainsEnd = lineText.contains("}");
+            if (isContainsEnd) {
+                dependenciesExists = false;
+            }
+        }
+    }
+
+    /**
+     * The detectMavenDependency method gets through the current open maven file and test the dependency statements to see if they are in the database.
+     * This will highlight targeted libraries which are in the database.
+     * @param editor is currently focused editor instance.
+     * @param psiFile is a PSI (Program Structure Interface) file, the root of a structure representing a file's contents as a hierarchy of elements.
+     * @param projectName is the name of current project.
+     */
+    public void detectMavenDependency(@NotNull final Editor editor, @NotNull final PsiFile psiFile, @NotNull final  String projectName ) throws IOException, SAXException {
+
+        final MarkupModel editorModel = editor.getMarkupModel();
+        final Document document = editor.getDocument();
+        TextAttributes attributes = EditorColorsManager.getInstance().getGlobalScheme().getAttributes(EditorColors.WRITE_SEARCH_RESULT_ATTRIBUTES);
+        TextAttributes softerAttributes = attributes.clone();
+
+        XmlFile xmlFile = (XmlFile) psiFile;
+        InputStream is = new ByteArrayInputStream(xmlFile.getText().getBytes());
+        org.w3c.dom.Document doc = PositionalXMLReader.readXML(is);
+        is.close();
+
+        int nodeIndex = 0;
+        NodeList nodeList = doc.getElementsByTagName("groupId");
+
+        MavenXpp3Reader Xpp3Reader = new MavenXpp3Reader();
+        editorModel.removeAllHighlighters();
+        VirtualFile vFile = psiFile.getOriginalFile().getVirtualFile();
+        String path = vFile.getPath();
+        mavenDepList.clear();
+        int currentLine;
+        try {
+            Model model = Xpp3Reader.read(new FileReader(path));
+            List<Dependency> dependencies = model.getDependencies();
+            String selectedTerm = null;
+
+            for (Dependency dependency : dependencies) {
+                selectedTerm = dependency.getGroupId();
+
+                DatabaseAccess dataAccessObject = new DatabaseAccess();
```

Comment:
do you need a new DatabaseAccess instance inside each loop?

Reply:
move out to the loop